pkg/controller/deployment/util/deployment_util.go

```diff
@@ -261,14 +261,25 @@ func SetNewReplicaSetAnnotations(deployment *apps.Deployment, newRS *apps.Replic
 	// If a revision annotation already existed and this replica set was updated with a new revision
 	// then that means we are rolling back to this replica set. We need to preserve the old revisions
 	// for historical information.
-	if ok && annotationChanged {
+	if ok && oldRevisionInt < newRevisionInt {
```

Comment:
Why does this needs to be changed?

Reply:
Previously annotationChanged may be true after this assignment:
```
	annotationChanged := copyDeploymentAnnotationsToReplicaSet(deployment, newRS)
```
I narrowed the condition to when new revision carries different value than the old rev.